pagination/pagination_test.go

```diff
@@ -0,0 +1,99 @@
+package pagination
+
+import (
+	. "github.com/smartystreets/goconvey/convey"
+	"net/http/httptest"
+	"testing"
+)
+
+const (
+	defaultLimit        = 10
+	defaultOffset       = 1
+	defaultMaximumLimit = 100
+)
+
+func TestReadPaginationValues(t *testing.T) {
+	defaultPaginator := NewPaginator(defaultLimit, defaultOffset, defaultMaximumLimit)
+
+	Convey("Given a request without pagination parameters", t, func() {
+		r := httptest.NewRequest("GET", "/endpoint_to_paginate", nil)
+		Convey("When SetPaginationValues is called", func() {
+
+			offset, limit, err := defaultPaginator.SetPaginationValues(r)
+			Convey("Then the default values are returned", func() {
+				So(err, ShouldBeNil)
+				So(limit, ShouldEqual, defaultLimit)
+				So(offset, ShouldEqual, defaultOffset)
+			})
+		})
+	})
+
+	Convey("Given a request with valid pagination parameters", t, func() {
+		r := httptest.NewRequest("GET", "/endpoint_to_paginate?limit=13&offset=2", nil)
+		Convey("When SetPaginationValues is called", func() {
+			offset, limit, err := defaultPaginator.SetPaginationValues(r)
+			Convey("Then the default values are overwritten by the ones in the request", func() {
+				So(err, ShouldBeNil)
+				So(limit, ShouldEqual, 13)
+				So(offset, ShouldEqual, 2)
+			})
+		})
+	})
+
+	Convey("Given a request with a negative limit value", t, func() {
+		r := httptest.NewRequest("GET", "/endpoint_to_paginate?limit=-13&offset=2", nil)
+		Convey("When SetPaginationValues is called", func() {
+			offset, limit, err := defaultPaginator.SetPaginationValues(r)
+			Convey("Then the an error is returned saying the limit value is invalid", func() {
+				So(err, ShouldBeError)
+				So(err, ShouldEqual, ErrInvalidLimitParameter)
+				So(limit, ShouldEqual, 0)
+				So(offset, ShouldEqual, 0)
+			})
+		})
+	})
+
+	Convey("Given a request with a negative offset value", t, func() {
+		r := httptest.NewRequest("GET", "/endpoint_to_paginate?limit=13&offset=-2", nil)
+		Convey("When SetPaginationValues is called", func() {
+			offset, limit, err := defaultPaginator.SetPaginationValues(r)
+			Convey("Then the an error is returned saying the offset value is invalid", func() {
+				So(err, ShouldBeError)
+				So(err, ShouldEqual, ErrInvalidOffsetParameter)
+				So(limit, ShouldEqual, 0)
+				So(offset, ShouldEqual, 0)
+			})
+		})
+	})
+
+	Convey("Given a request with a limit value that exceeds the default maximum limit", t, func() {
+		r := httptest.NewRequest("GET", "/endpoint_to_paginate?limit=101&offset=2", nil)
+		Convey("When SetPaginationValues is called", func() {
+			offset, limit, err := defaultPaginator.SetPaginationValues(r)
+			Convey("Then the an error is returned saying the limit value is invalid and is over the maximum value", func() {
+				So(err, ShouldBeError)
+				So(err, ShouldEqual, ErrLimitOverMax)
+				So(limit, ShouldEqual, 0)
+				So(offset, ShouldEqual, 0)
+			})
+		})
+	})
+
```

Comment:
Think about adding a couple more tests to capture when offset and limit are set to non-numerical value:

`?offset=two` and `?limit=ten`